sources/brush/src/main.rs

```diff
@@ -0,0 +1,429 @@
+    let program = program.as_ref();
+    let paths = paths.as_ref();
+    let cwd = std::env::current_dir().unwrap_or_default();
+    which::CanonicalPath::new_in(program, Some(paths), cwd)
+        .with_whatever_context(|_| format!("failed to locate '{}' in PATH", program.display()))
+}
+
+// Find the canonical path to the provided symlink, resolving any intermediate symlinks or relative
+// path components.
+fn resolve_symlink(symlink: impl AsRef<Path>) -> Result<PathBuf> {
+    let symlink = symlink.as_ref();
+    fs::canonicalize(symlink)
+        .with_whatever_context(|_| format!("failed to canonicalize '{}'", symlink.display()))
+}
+
+// Helper function to extract the filename for the given path.
+fn filename(path: impl AsRef<Path>) -> Result<OsString> {
+    let path = path.as_ref();
+    let path = path
+        .file_name()
+        .with_whatever_context(|| format!("failed to get file name for '{}'", path.display()))?
+        .to_os_string();
+    Ok(path)
+}
+
+// Verify that the program is allowed. If successful, it will return the canonical path to the
+// allowed program, along with the filename that should be used as arg0. This allows multi-call
+// binaries such as `coreutils` to know which personality to use upon execution.
+fn verify_program(
+    allowed_programs_dir: impl AsRef<OsStr>,
+    allowed_paths: impl AsRef<str>,
+    program: impl AsRef<Path>,
+) -> Result<(PathBuf, OsString)> {
+    let requested_program_path = program.as_ref();
+    let requested_program_name = filename(requested_program_path)?;
+    let allowed_program_path: PathBuf = [allowed_programs_dir.as_ref(), &requested_program_name]
+        .iter()
+        .collect();
+
+    let resolved_program_path = resolve_program(requested_program_path, allowed_paths)
+        .with_whatever_context(|_| {
+            format!(
+                "'{}' could not be found",
+                requested_program_name.to_string_lossy()
+            )
+        })?;
+
+    let verified_program_path =
+        resolve_symlink(&allowed_program_path).with_whatever_context(|_| {
+            format!(
+                "'{}' is not allowed",
+                requested_program_name.to_string_lossy()
+            )
+        })?;
+
+    ensure_whatever!(
+        resolved_program_path == verified_program_path,
+        "found symlink '{}' for '{}', but it resolves to '{}', not '{}'",
+        allowed_program_path.to_string_lossy(),
+        requested_program_name.to_string_lossy(),
+        verified_program_path.to_string_lossy(),
+        resolved_program_path.to_string_lossy(),
+    );
+
+    Ok((verified_program_path, requested_program_name))
+}
+
+// Wrapper function that either executes the requested program, if allowed, or returns an error.
+fn run(args: Vec<String>, allowed_programs_dir: &str, allowed_paths: &str) -> Result<()> {
+    let (program, args) = parse_args(args)?;
+    let (program, arg0) = verify_program(allowed_programs_dir, allowed_paths, &program)?;
```

Comment:
nit: although more efficient to reuse the same variable, I think I'd feel a bit better about keeping the unverified `program` in a separate variable to avoid future refactoring that loses the safety of only executing the verified program. 
```suggestion
    let (verified_program, arg0) = verify_program(allowed_programs_dir, allowed_paths, &program)?;
```

Reply:
> nit: although more efficient to reuse the same variable

Rust uses SSA so there is no efficiency gain or loss either way

> I'd feel a bit better about keeping the unverified program in a separate variable to avoid future refactoring that loses the safety of only executing the verified program.

I didn't want the "old" binding for `program` here to stay alive, but having `verify_program` take an owned value for `program` is annoying for unit tests.